File: CSharp/Studies/Studies/ArrayUtils.cs
using System;

namespace Studies
{
    public static class ArrayUtils
    {
        /// <summary>
        /// https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/727/
        /// </summary>
        /// <returns>The length of the resized array</returns>
        /// <param name="arr">an array of sorted ints.</param>
        public static int RemoveDuplicatesFromSortedInts(ref int[] arr)
        {
            var shiftCount = 0;
            var shiftedArrLength = arr.Length;

            for(var i = 1; i < shiftedArrLength; i++)
            {
                if (arr[i] == arr[i-1])
                {
                    i--;
                    shiftCount++;
                    shiftedArrLength--;
                    Array.Copy(arr, i + 1, arr, i, arr.Length - (i + 1));
                }
            }

            Array.Resize(ref arr, shiftedArrLength);

            return arr.Length;
        }


        /// <summary>
        /// https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/564/
        /// </summary>
        /// <param name="prices">Prices of various stocks indexed by day</param>
        /// <returns>The max profit derived from buying and selling stock from <paramref name="prices"/></returns>
        public static int MaxProfit(int[] prices)
        {
            // for an array [a, b, c, d]
            // (b - a) + (d - c) > (d - a)
            // only when b > c

            var buying = false;
            var profit = 0;
            var purchased = 0;
            var maxProfit = 0;

            for (var i = 0; i < prices.Length - 1; i++)
            {
                var potentialProfit = prices[i + 1] - prices[i];

                if (potentialProfit > 0)
                {
                    purchased = buying ? purchased : prices[i];
                    profit = prices[i + 1] - purchased;
                    buying = true;
                }
                else
                {
                    maxProfit += profit;
                    buying = false;
                }
            }

            if (buying) maxProfit += profit;

            return maxProfit;
        }


    }
}

File: js/lib/arrays/ArrayUtils.js
var Queue = require("../data_structures/Queue");

class ArrayUtils {
    /**
     * 
     * @param {Array} ints - a sorted set of numbers to remove duplicates from
     * @summary https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/727/
     * 
     * @returns new length of array after duplicates have been removed
     */
    static removeDuplicatesFromSortedInts(ints) {        
        for (let i = 1; i < ints.length; i ++) {
            if (ints[i] == ints[i-1]) {
                ints.splice(i, 1);
                i--;
            }
        }

        return ints.length;
    }

    /**
     * 
     * @param {Array} prices - Prices of various stocks indexed by day
     * @summary https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/564/
     * 
     * @returns The max profit derived from buying and selling stock from {prices}
     */
    static maxProfit(prices) {
        if (prices === undefined || !Array.isArray(prices)) return 0;

        let buying = false;
        let profit = 0;
        let purchased = 0;
        let maxProfit = 0; 

        for(let i = 0; i < prices.length - 1; i++) {
            var potentialProfit = prices[i + 1] - prices[i];

            if (potentialProfit > 0) {
                purchased = buying ? purchased : prices[i];
                profit = prices[i + 1] - purchased;
                buying = true;
            } else {
                maxProfit += profit;
                buying = false;
            }
        }

        if (buying) maxProfit += profit;

        return maxProfit;
    }

    /**
     * Rotates an array by k with O(1) space
     * https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/646/
     * @param {Array} nums An array of numbers
     * @param {rotate} k The number of slots to rotate each number
     */
    static rotate(nums, k) {
        let shift = k % nums.length;
        
        let q = new Queue();
        
        let lmt = 1;

        for (let i = 0; i < lmt; i++) {
            let j = i;
            let cnt = 0;
            q.enqueue(nums[i]);

            do {
                cnt++;
                j += shift;
                if (j >= nums.length) j -= nums.length;
                if (j !== i)  q.enqueue(nums[j]);
                nums[j] = q.dequeue();
            }
            while (j !== i);

            if (lmt === 1) lmt = nums.length / cnt;
        }
    }
}

module.exports = ArrayUtils;